java/src/main/java/ai/rapids/cudf/ColumnView.java

```diff
@@ -2531,12 +2531,37 @@ public final ColumnVector stringLocate(Scalar substring, int start, int end) {
    *                     regular expression pattern or just by a string literal delimiter.
    * @return list of strings columns as a table.
    */
+  @Deprecated
   public final Table stringSplit(String pattern, int limit, boolean splitByRegex) {
+    if (splitByRegex) {
+      RegexProgram regexProg = new RegexProgram(pattern, CaptureGroups.NON_CAPTURE);
+      return stringSplit(regexProg, limit);
+    } else {
+      return stringSplit(pattern, limit);
+    }
+  }
+
+  /**
+   * Returns a list of columns by splitting each string using the specified regex program. The
+   * number of rows in the output columns will be the same as the input column. Null entries
+   * are added for a row where split results have been exhausted. Null input entries result in
+   * all nulls in the corresponding rows of the output columns.
+   *
+   * @param regexProg the regex program with UTF-8 encoded string identifying the split pattern
+   *                  for each input string.
+   * @param limit the maximum size of the list resulting from splitting each input string,
+   *              or -1 for all possible splits. Note that limit = 0 (all possible splits without
+   *              trailing empty strings) and limit = 1 (no split at all) are not supported.
+   * @return list of strings columns as a table.
+   */
+  public final Table stringSplit(RegexProgram regexProg, int limit) {
     assert type.equals(DType.STRING) : "column type must be a String";
-    assert pattern != null : "pattern is null";
-    assert pattern.length() > 0 : "empty pattern is not supported";
+    assert regexProg != null : "regex program is null";
+    assert regexProg.pattern() != null : "pattern is null";
+    assert regexProg.pattern().length() > 0 : "empty pattern is not supported";
```

Comment:
nit: Do we ever want to allow a program to have an empty pattern or null pattern. Should these just be part of RegexProgram instead?

Reply:
Thank you! Do you think it's a good idea to conform with the C++ regex program implementation expectation?

Reply:
If C++ is already doing checks for us, then we don't need to replicate them here. If not and we know that they are going to cause problems, then we should add the checks here.

Reply:
I added pattern null check in RegexProgram constructor. I saw `cudf::strings::regex_program::create(empty_pattern)` in C++ tests so I think empty pattern should be allowed. Please let me know if I am wrong. Thanks!

Reply:
Sorry I just found the reason for this.

https://github.com/rapidsai/cudf/blob/ed6daad3a7edca6b8873b1105a585e12757dcf1c/java/src/main/native/src/ColumnViewJni.cpp#L703-L707

It looks like because the behavior is different from java we are not allowing it.  The point of the CUDF API is not to match java behavior but to expose the CUDF API.  Can we remove the check here and in the native code too. We will keep it in the plugin because that is where we care about it being the same.

(And do it for stringSplitRecord too)

This is still a nit so I am happy to let this in as is, but we could just file a follow on issue to fix it if you want.

Reply:
I see. Thank you very much for clarifying this! I can remove them now. To confirm, we want to remove the check from the ColumnView.java stringSplit APIs as well?